[cxx-interop] Add initial documentation for strict memory safety (#980)
swiftlang/swift-org-website: documentation/cxx-interop/safe-interop/index.md

---
layout: page
title: Mixing Strict Memory Safe Swift and C++
official_url: https://swift.org/documentation/cxx-interop/safe-interop/
redirect_from:
- /documentation/cxx-interop/safe-interop.html
---

## Table of Contents
{:.no_toc}

* TOC
{:toc}

## Introduction

> Swift 6.2 introduces [strict memory safety mode](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0458-strict-memory-safety.md),

Reviewer: Aren't the safe overloads we provide completely separate from the "strict safe mode"? I've been using them without having this mode on. Introducing the interop makes it seems like it's coupled to "strict safe mode" and I don't think that's the case.

an opt-in language feature that prevents common memory safety issues by running
extra compile-time safety checks and emitting diagnostics.

This document describes the way Swift/C++ interoperability works when strict

> safe mode is enabled in Swift.

Reviewer: Aren't the safe overloads we provide completely separate from the "strict safe mode"? I've been using them without having this mode on. Introducing the interop makes it seems like it's coupled to "strict safe mode" and I don't think that's the case.

Future releases of Swift might change how Swift and C++
interoperate,
as the Swift community gathers feedback from real world adoption of C++
interoperability in mixed Swift and C++ codebases.
Please provide the feedback that you have on the
[Swift Forums](https://forums.swift.org/c/development/c-interoperability/), or
by filing an [issue on GitHub](https://github.com/swiftlang/swift/issues/new/choose).
Future changes to the design or functionality of C++ interoperability will not
break code in existing codebases [by default](#source-stability-guarantees-for-mixed-language-codebases).
</div>

## Overview

Swift provides memory safety with a combination of language affordances and runtime checking.
However, Swift also deliberately includes some unsafe constructs, such as the `UnsafePointer` and `UnsafeMutablePointer`
types in the standard library.
In some cases, Swift needs additional information that is not present in the C++ type and API declarations
to safely interface with them. This document describes how such code needs to be annotated.

### Annotating Foreign Types

Types imported from C++ are considered foreign to Swift. Normally, most C++ types are imported into Swift

> without any restriction. However, a small set of C++ APIs e.g. pointers/references and methods returning

Author: Good point, rephrased it.

> pointers will be imported as unsafe (section [Working with C++ references and view types in Swift](https://www.swift.org/documentation/cxx-interop/#working-with-c-references-and-view-types-in-swift)

Reviewer 2: It should have been rendered as an emdash, i.e., "—", by Markdown... but now I realize that isn't standard Markdown (though it is supported by some extensions).

I've updated the suggestion to just a unicode emdash.

```

Let's try to use them from Swift with strict memory safety enabled:

```swift
func getFileName(_ path: borrowing std.string) -> StringRef {
    let normalizedPath = normalize(path)
    return fileName(normalizedPath)
}
```

Building this code will emit a warning that the `fileName` call is unsafe because
it references the unsafe type `StringRef`. Swift considers `StringRef` unsafe because
it has a pointer member. Pointers in types like `StringRef` can dangle, so we need to take extra
care that the buffer they point to outlives the `StringRef` object they are encapsulated by.

Swift's [non-escapable types](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0446-non-escapable.md)
can also have lifetime dependencies, just like `StringRef`. However, the Swift compiler
can track these dependencies and enforce safety at compile time. To import `StringRef`
as a safe type, we need to mark it as a non-escapable type by annotating the class
definition:

```c++
class SWIFT_NONESCAPABLE StringRef { ... };
```

Now the Swift compiler imports `StringRef` as a safe type and no longer
emits a warning about using an unsafe type.

> ### Annotating C++ APIs

Reviewer: There's no discussion of `__counted_by`, `__counted_by_or_null`, `__sized_by`, or `__sized_by_or_null`, or `__noescape` here. I feel like these should be mentioned.

Building the code again will emit a new diagnostic for the `fileName` function about
missing lifetime annotations. Functions returning non-escapable types need annotations

> to describe their lifetime contracts via [lifetimebound](https://clang.llvm.org/docs/AttributeReference.html#id11)

Author: Thanks for noticing this!

```swift
func getFileName(_ path: borrowing std.string) -> StringRef {
    let normalizedPath = normalize(path)
    // error: lifetime-dependent value escapes local scope
    // note: depends on `normalizedPath`
    return fileName(normalizedPath)
}
```

The value returned by `fileName` will dangle after the lifetime of `normalizedPath` ends.
We can fix this error by pushing the task of normalizing a path to the callee:

```swift
// Path needs to be normalized.
func getFileName(_ path: borrowing std.string) -> StringRef {
    return fileName(normalizedPath)
}
```

Or we could return an `Escapable` value like `std.string` instead of a dangling `StringRef`:

```swift
func getFileName(_ path: borrowing std.string) -> std.string {
    let normalizedPath = normalize(path)
    let ref = fileName(normalizedPath)
    return ref.toString()
}
```

After annotating the C++ code, the Swift compiler can enforce the lifetime
contracts helping us to write code that is free of memory safety errors.

## Escapability Annotations in Detail

Under the strictly safe mode, even though compiler warns on unannotated types,
they are imported as if they are `Escapable` to maintain backward

> compatibility. This might change in the future under a new interoperability version.

Author: Thanks!

We have already seen that we can import a type as `~Escapable` to Swift by adding
the `SWIFT_NONESCAPABLE` annotation:

```c++
struct SWIFT_NONESCAPABLE View {
    View(const int *p) : member(p) {}
private:
    const int *member;

> };

Author: clang-format and Xcode seem to default to `type *ptr`, but I agree that we should make it consistent across the document.

```

Moreover, we can explicitly mark types as `Escapable` using the `SWIFT_ESCAPABLE`
annotation to express that they are not lifetime dependent on any other values:

```c++

> struct SWIFT_ESCAPABLE Owner { ... }; 

Author: Made these consistent across the document 👍

with an `Escapable` type.

The `SWIFT_ESCAPABLE_IF` macro can take multiple template parameters:

```c++
template<typename F, typename S>
struct SWIFT_ESCAPABLE_IF(F, S) MyPair {
    F first;
    S second;
};
```

`MyPair` instantiations are only imported as `Escapable` if both template arguments
are `Escapable`.

`Escapable` types cannot have `~Escapable` fields. The following code snippet will
trigger a compiler error:

```c++
struct SWIFT_NONESCAPABLE View { ... };
struct SWIFT_ESCAPABLE Owner { 
    View v;
}; 
```

Escapability annotations will not only help the Swift compiler to import C++ types
safely, it will also help discover missing lifetime annotations as all `~Escapable`
parameters and return values need to be annotated in an API to make its use safe in
Swift.

## Lifetime Annotations in Detail

The `lifetimebound` attribute on a function parameter or implicit object parameter
indicates that the returned object's lifetime could end when any of the `lifetimebound`
annotated parameters' lifetime ended.
This annotation a constructor describes the lifetime of the created object:

```c++
struct SWIFT_NONESCAPABLE View {

>     View(const int *p [[clang::lifetimebound]]) : member(p) {}

Author: Made these consistent across the document 👍

    ...
};
```

In this example, the object initialized by the `View` constructor has the same
lifetime as the input argument of the constructor.

In case the attribute is after the method signature, the returned object has
the same lifetime as the implicit `this` parameter.

```c++
struct Owner {
    int data;

    View handOutView() const [[clang::lifetimebound]] {
        return View(&data);
    }
};
```

Consider a call site like `View v = o.handOutView()`. The `v` object has the same lifetime

> as `o`.

Author: Done

In case the attribute is applied to a subset of the parameters, the return
value might depend on the corresponding arguments:

```c++
View getOneOfTheViews(const Owner& owner1 [[clang::lifetimebound]], const Owner& owner2
    View view1 [[clang::lifetimebound]], View view2 [[clang::lifetimebound]]) {
    if (coinFlip)
        return View(&owner1.data);
    if (coinFlip)
        return view1;
    else

>         return view2;

Author: Yeah, I would expect `coinFlip()` to be non-deterministic based on the name, but I don't feel strongly here.

}
```

Here, the returned `View`'s lifetime depends on `owner`, `view1`, and `view2` but it cannot
depend on `owner2`.

Occasionally, a function might return a non-escapable type that has no dependency on any other values.
These types might point to static data or might represent an empty sequence or lack of data.
Such functions need to be annotated with `SWIFT_RETURNS_INDEPENDENT_VALUE`:

```c++
View returnsEmpty() SWIFT_RETURNS_INDEPENDENT_VALUE {
    return View();
}
```

Notably, the default constructor of a type is always assumed to create an independent value.

We can also annotate `lifetimebound` APIs via APINotes. The `-1` index represents the `this` position.

```
Tags:
- Name: MyClass
  Methods:
    - Name: annotateThis
      Parameters:
        - Position:      -1
          Lifetimebound: true
    - Name: methodToAnnotate
      Parameters:
        - Position:      0
          Lifetimebound: true
```

> Note that APINotes have some limitations around C++, they do not support overloaded functions.

Author: Agreed on the spacing in `API Notes`, but would it be more natural to say `... notes have ...` rather than `... notes has ...`?

While `lifetimebound` always describes the lifetime dependencies of the return value (or
the constructed object in case of constructors), we can use can use `lifetime_capture_by`
annotation to describe the lifetime of other output values, like output/inout arguments
or globals.

```c++
void copyView(View view1 [[clang::lifetime_capture_by(view2)]], View &view2) {
    view2 = view1;
}
```

In this example, `view2` will have get all of the lifetime dependencies of `view1`
after a call to `copyView`. a

We can annotate dependency captured by the implicit `this` object, or
an inout argument capturing `this`:

```c++
struct SWIFT_NONESCAPABLE CaptureView {
    void captureView(View v [[clang::lifetime_capture_by(this)]]) {
        view = v;
    }

    void handOut(View &v) const [[clang::lifetime_capture_by(v)]] {
       v = view; 
    }

    View view;

> };

Author: Done, renamed the field to `containedView`.

```

All of the non-escapable inputs need lifetime annotations for a function to be
considered safe. If an input never escapes from the called function we can use
the `noescape` annotation:

```c++
void is_palindrome(std::span<int> s [[clang::noescape]]);
```

While the annotations in this section are powerful, they cannot express all of
the lifetime contracts. APIs with inexpressible contracts can be used from Swift,
but they are imported as unsafe APIs and need extra care from the developers
to manually guarantee safety.

## Convenience Overloads for Annotated Spans and Pointers

> C++ APIs often using standard library types or other constructs like a
> pointer and a size to represent buffers that have Swift equivalents like
> Swift's `Span` type. These Swift types have additional requirements and
> guarantees. When these properties are properly annotated on the C++ side,
> the Swift compiler can introduce safe convenience functions to make
> interacting with the C++ APIs as effortless as if they were written in Swift.

Reviewer 2: ```suggestion
C++ APIs often feature parameters that denote a span of memory.
For example, some might have two parameters where one points to a memory buffer and the other designates the buffer's size;
others might use the [`std::span`](https://en.cppreference.com/w/cpp/container/span) type from the C++ standard library.
When such APIs are appropriately annotated, the Swift compiler can bridge those span-like parameters to Swift's [`Span`](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0447-span-access-shared-contiguous-storage.md) and [`MutableSpan`](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0467-MutableSpan.md) types, and the user with a safe and convenient interface to those imported APIs.
```

Reviewer 2: We are missing documentation for parameters annotated with `__counted_by` and `__sized_by`.

Reviewer 2: If that's only coming in a follow-up patch, then maybe we shouldn't even mention the case where we denote a span with two separate parameters, where one is a pointer and the other is a size/length.

> ### C++ `std::span` Support

Reviewer: I don't really know where to fit this in but there are no examples of using `__counted_by` and friends here. I realize they are not as safe as span but we are providing overloads that are safer that their original counter parts.

APIs taking or returning C++'s `std::span` with sufficient lifetime

> annotations will automatically get overloads taking/returning Swift
> `Span`.

Reviewer 2: ```suggestion
annotations will automatically get safe overloads take or return Swift
`Span` and `MutableSpan` types.
```

The following table summarizes the generated convenience overloads:

```c++
using IntSpan = std::span<const int>;
using IntVec = std::vector<int>;
```

| C++ API                                                   | Generated Swift overload                                             |
| --------------------------------------------------------- | -------------------------------------------------------------------- |
| `void takeSpan(IntSpan x [[clang::noescape]]);`           | `func takeSpan(_ x: Span<Int32>)`                                    |
| `IntSpan changeSpan(IntSpan x [[clang::lifetimebound]]);` | `@lifetime(x) func changeSpan(_ x: Span<Int32>) -> Span<Int32>`      |
| `IntSpan changeSpan(IntVec& x [[clang::lifetimebound]]);` | `@lifetime(x) func changeSpan(_ x: borrowing IntVec) -> Span<Int32>` |
| `IntSpan Owner::getSpan() [[clang::lifetimebound]];`      | `@lifetime(self) func getSpan() -> Span<Int32>`                      |

> These transformations only support top level `std::span`s. The compiler

Reviewer 2: ```suggestion
These transformations only support top-level `std::span`s. The compiler
```

> currently does not transform nested `std::span`s. A `std::span` of a non-const

Reviewer 2: ```suggestion
currently does not transform nested `std::span`s. A `std::span<T>` of a non-const
```

> type `T` is transformed to `MutableSpan<T>` on the Swift wide.

Reviewer 2: ```suggestion
type `T` is transformed to `MutableSpan<T>` on the Swift side.
```

---
layout: page

> title: Mixing Strict Memory Safe Swift and C++

Reviewer 3: Suggested title: "Safely mixing C, C++, and Swift" (We need a follow up to add "C" interop part).

Author: We don't really talk about the C part at this point, so I think we can call this "Safely Mixing Swift and C++". I amended the title.

official_url: https://swift.org/documentation/cxx-interop/safe-interop/
redirect_from:
- /documentation/cxx-interop/safe-interop.html
---

## Table of Contents
{:.no_toc}

* TOC
{:toc}

## Introduction

This document describes the additional memory safety features that Swift
provides when interoperating with C and C++. It describes the ways to make Swift
aware of certain memory safety aspects of your codebase, as well as how Swift
uses that knowledge to prevent common memory safety issues in your code.

* * *

<div class="info" markdown="1">
C++ interoperability is an actively evolving feature of Swift.
Future releases of Swift might change how Swift and C++
interoperate,
as the Swift community gathers feedback from real world adoption of C++
interoperability in mixed Swift and C++ codebases.
Please provide the feedback that you have on the
[Swift Forums](https://forums.swift.org/c/development/c-interoperability/), or
by filing an [issue on GitHub](https://github.com/swiftlang/swift/issues/new/choose).
Future changes to the design or functionality of C++ interoperability will not

> break code in existing codebases [by default](#source-stability-guarantees-for-mixed-language-codebases).

Reviewer 4: Not sure if this will be added later, but I noticed that this link is dangling, just so we don't forget to write this part
